bip-0360.mediawiki

```diff
@@ -170,386 +172,269 @@ using Grover's algorithm would require at least 10^24 quantum operations. As for
+and tapscript code already in Bitcoin. This reduces the implementation burden on wallets, exchanges, and 
+libraries since they can reuse code they already have.
 
-  0x01 | 0x02 | 0x04 | 0x08 = 0x0F
+By separating P2QRH from the introduction of PQ signatures relying parties can move from P2TR to P2QRH 
+without simultaneously having to change from Schnorr signatures to PQ signatures. Simply moving coins from 
+P2TR to P2QRH protects those coins from long-exposure quantum attacks. Then to gain full quantum resistance, 
+verification of PQ signatures can be added as an additional tapleaf alongside Schnorr signatures<ref name="mattleaf">Inspired by Matt Corallo's discussion of this idea on the Bitcoin-dev mailing list [https://groups.google.com/g/bitcoindev/c/oQKezDOc4us/m/T1vSMkZNAAAJ Re: P2QRH / BIP-360 Update]</ref>. 
+When quantum attacks become practical, users would then be fully protected as the P2QRH output would allow
+them to switch to sending their coins using the PQ signature algorithms. This allows the upgrade to quantum 
+resistance to be largely invisible to users.
 
-==== Hash Commitment ====
+Consider the P2QRH output with three tapscripts:
 
-If there is only a single public key, the hash is computed as the HASH256 of the public key.
+* Spend requires a Schnorr signature
+* Spend requires a ML-DSA signature
+* Spend requires a SLH-DSA signature
 
-In order to support multiple keys, as in the context of multisig or singlesig hybrid cryptography, the hash is
-computed as a commitment to a vector of public key hashes:
+In the event that Schnorr signatures are broken, users can spend their coins using ML-DSA. 
+If both Schnorr and ML-DSA are broken, the user can still rely on SLH-DSA.
 
-1. Sort the public keys first by key type, then by public key value
-2. For each sorted public key, compute its HASH256
-3. Concatenate all the public key hashes in sorted order
-4. Prepend key type bitmask and threshold to the concatenated hashes
-5. Compute the HASH256 of the result
+One intent in supporting Schnorr, ML-DSA, and SLH-DSA in tapscript, is to allow parties to construct outputs such that funds 
+are still secure even if two of the three the signature algorithms are completely broken. This is motivated by the use case
+of securely storing Bitcoins in a cold wallet for very long periods of time (50 to 100 years).
 
-For example with 4 public keys:
+For PQ signatures we considered the NIST approved SLH-DSA (SPHINCS+), ML-DSA (CRYSTALS-Dilithium), 
+FN-DSA (FALCON). Of these three algorithms, SLH-DSA has the largest signature size, but is the most conservative 
+choice from a security perspective because SLH-DSA is based on well studied and time-tested hash-based cryptography. 
+Both FN-DSA and ML-DSA signatures are significantly smaller than SLH-DSA signatures but are based on newer lattice-based 
+cryptography. Since ML-DSA and FN-DSA are both similar lattice-based designs, we choose to only support one of them as the 
```

Comment:
While it’s true that an advance in lattice based cryptanalysis might break both ML-DSA and FN-DSA, the constructions are sufficiently different that it is quite possible that one might be broken while the other remains intact. I’d argue for the inclusion of both.

Reply:
If we choose to include both, I can make a PR against Ethan’s fork to fix the other instances where only ML-DSA is specified.